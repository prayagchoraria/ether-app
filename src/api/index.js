import { ethers } from 'ethers';

export const fetchBlocks = async () => {
  const provider = new ethers.providers.Web3Provider(window.ethereum, 'any');
  const latestBlockNumber = await provider.getBlockNumber();

  const blocks = [];
  for (let i = 0; i < 10; i += 1) {
    const block = provider.getBlock(latestBlockNumber - i);
    blocks.push(block);
  }
  return Promise.all(blocks);
};
