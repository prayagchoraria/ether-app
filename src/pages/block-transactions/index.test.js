import React from 'react';
import { useParams } from 'react-router-dom';
import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react';
import { fetchBlockTransactions } from '../../api';
import BlockTransactions from '.';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(),
}));

jest.mock('../../api', () => ({
  fetchBlockTransactions: jest.fn(),
}));

const renderBlockTransactions = () => {
  const transactions = [
    {
      hash: '0x395950bd3b8e',
      gasLimit: { _hex: '0x01ca35d2' },
      gasPrice: { _hex: '0x12564a62b2' },
      nonce: 12345,
      from: '0xA7EFAe72',
      to: '0xDd52Db7e',
      value: { _hex: '0x00' },
    },
    {
      hash: '0x5182e0cd0bd8',
      gasLimit: { _hex: '0x01ba35e2' },
      gasPrice: { _hex: '0x12564b62a2' },
      nonce: 23456,
      from: '0xAe45a8240',
      to: '0x7fA4737Ea',
      value: { _hex: '0x277df16f52c000' },
    },
  ];
  useParams.mockReturnValue({ blockNumber: 123 });
  fetchBlockTransactions.mockReturnValue(Promise.resolve({ transactions }));

  const renderedBlockTransactions = render(<BlockTransactions />);
  return renderedBlockTransactions;
};

describe.only('BlockTransactions Component', () => {
  it('displays the header', async () => {
    useParams.mockReturnValue({ blockNumber: 123 });
    const { getByText } = render(<BlockTransactions />);
    expect(useParams).toHaveBeenCalled();
    expect(getByText('Transactions of Block 123')).toBeInTheDocument();
  });

  it('renders loading state', () => {
    useParams.mockReturnValue({ blockNumber: 123 });
    const { getByText } = render(<BlockTransactions />);
    expect(useParams).toHaveBeenCalled();
    expect(getByText('Loading...')).toBeInTheDocument();
  });

  it('renders error state', async () => {
    useParams.mockReturnValue({ blockNumber: 123 });
    fetchBlockTransactions.mockReturnValue(
      Promise.reject(new Error('Something went wrong!'))
    );
    const { queryByText } = render(<BlockTransactions />);
    expect(fetchBlockTransactions).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(queryByText('Loading...')).not.toBeInTheDocument();
    });
    await waitFor(() => {
      expect(queryByText('Error: Something went wrong!')).toBeInTheDocument();
    });
  });

  it('matches loading snapshot', async () => {
    const tree = renderBlockTransactions();
    expect(tree).toMatchSnapshot();
  });

  it('matches error snapshot', async () => {
    useParams.mockReturnValue({ blockNumber: '123' });
    fetchBlockTransactions.mockReturnValue(
      Promise.reject(new Error('Something went wrong!'))
    );
    const tree = render(<BlockTransactions />);
    await waitFor(() => {
      expect(tree.queryByText('Loading...')).not.toBeInTheDocument();
    });
    expect(tree).toMatchSnapshot();
  });

  it('only render transactions with value', async () => {
    const tree = renderBlockTransactions();
    const { queryByText } = tree;
    await waitFor(() => {
      expect(fetchBlockTransactions).toHaveBeenCalledTimes(1);
    });
    await waitFor(() => {
      expect(queryByText('Loading...')).not.toBeInTheDocument();
    });
    await waitFor(() => {
      expect(queryByText('0x395950bd3b8e')).not.toBeInTheDocument();
      expect(queryByText('0x5182e0cd0bd8')).toBeInTheDocument();
    });
  });

  it('matches rendered snapshot', async () => {
    const tree = renderBlockTransactions();
    await waitFor(() => {
      expect(tree.queryByText('Loading...')).not.toBeInTheDocument();
    });
    expect(tree).toMatchSnapshot();
  });
});
