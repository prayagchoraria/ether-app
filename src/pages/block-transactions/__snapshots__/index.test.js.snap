// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockTransactions Component matches error snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="transactions"
      >
        <h2>
          Transactions of Block 
          123
        </h2>
        <span
          class="error"
        >
          Error: 
          Something went wrong!
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="transactions"
    >
      <h2>
        Transactions of Block 
        123
      </h2>
      <span
        class="error"
      >
        Error: 
        Something went wrong!
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BlockTransactions Component matches loading snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="transactions"
      >
        <h2>
          Transactions of Block 
          123
        </h2>
        <span>
          Loading...
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="transactions"
    >
      <h2>
        Transactions of Block 
        123
      </h2>
      <span>
        Loading...
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BlockTransactions Component matches rendered snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="transactions"
      >
        <h2>
          Transactions of Block 
          123
        </h2>
        <div
          class="search"
        >
          <span>
            Search address
          </span>
          <input
            placeholder="Search"
            type="text"
          />
        </div>
        <div
          class="data-container"
        >
          <div
            class="info-column"
          >
            <div>
              <span>
                Transaction Hash
              </span>
              <span>
                0x5182e0cd0bd8
              </span>
            </div>
            <div>
              <span>
                From
              </span>
              <span>
                0xAe45a8240
              </span>
            </div>
            <div>
              <span>
                To
              </span>
              <span>
                0x7fA4737Ea
              </span>
            </div>
          </div>
          <div
            class="info-row"
          >
            <div>
              <span>
                Gas Limit
              </span>
              <span>
                28980706
              </span>
            </div>
            <div>
              <span>
                Gas Price
              </span>
              <span>
                78757192354
              </span>
            </div>
            <div>
              <span>
                Value
              </span>
              <span>
                11116000000000000
              </span>
            </div>
            <div>
              <span>
                NONCE
              </span>
              <span>
                23456
              </span>
            </div>
          </div>
        </div>
        <div
          class="data-container"
        >
          <div
            class="info-column"
          >
            <div>
              <span>
                Transaction Hash
              </span>
              <span>
                0x5182e0cd1e58
              </span>
            </div>
            <div>
              <span>
                From
              </span>
              <span>
                0xA7EFAe72
              </span>
            </div>
            <div>
              <span>
                To
              </span>
              <span>
                0x7fA4737Ea
              </span>
            </div>
          </div>
          <div
            class="info-row"
          >
            <div>
              <span>
                Gas Limit
              </span>
              <span>
                28980706
              </span>
            </div>
            <div>
              <span>
                Gas Price
              </span>
              <span>
                78757192354
              </span>
            </div>
            <div>
              <span>
                Value
              </span>
              <span>
                11116000000000000
              </span>
            </div>
            <div>
              <span>
                NONCE
              </span>
              <span>
                23456
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="transactions"
    >
      <h2>
        Transactions of Block 
        123
      </h2>
      <div
        class="search"
      >
        <span>
          Search address
        </span>
        <input
          placeholder="Search"
          type="text"
        />
      </div>
      <div
        class="data-container"
      >
        <div
          class="info-column"
        >
          <div>
            <span>
              Transaction Hash
            </span>
            <span>
              0x5182e0cd0bd8
            </span>
          </div>
          <div>
            <span>
              From
            </span>
            <span>
              0xAe45a8240
            </span>
          </div>
          <div>
            <span>
              To
            </span>
            <span>
              0x7fA4737Ea
            </span>
          </div>
        </div>
        <div
          class="info-row"
        >
          <div>
            <span>
              Gas Limit
            </span>
            <span>
              28980706
            </span>
          </div>
          <div>
            <span>
              Gas Price
            </span>
            <span>
              78757192354
            </span>
          </div>
          <div>
            <span>
              Value
            </span>
            <span>
              11116000000000000
            </span>
          </div>
          <div>
            <span>
              NONCE
            </span>
            <span>
              23456
            </span>
          </div>
        </div>
      </div>
      <div
        class="data-container"
      >
        <div
          class="info-column"
        >
          <div>
            <span>
              Transaction Hash
            </span>
            <span>
              0x5182e0cd1e58
            </span>
          </div>
          <div>
            <span>
              From
            </span>
            <span>
              0xA7EFAe72
            </span>
          </div>
          <div>
            <span>
              To
            </span>
            <span>
              0x7fA4737Ea
            </span>
          </div>
        </div>
        <div
          class="info-row"
        >
          <div>
            <span>
              Gas Limit
            </span>
            <span>
              28980706
            </span>
          </div>
          <div>
            <span>
              Gas Price
            </span>
            <span>
              78757192354
            </span>
          </div>
          <div>
            <span>
              Value
            </span>
            <span>
              11116000000000000
            </span>
          </div>
          <div>
            <span>
              NONCE
            </span>
            <span>
              23456
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
